### ISSUE TITLE
False Positive: Benchmark-excluded string `==` cases are still mechanically treated as comparison mistakes.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6544]
# False Positive: Benchmark-excluded string `==` cases are still mechanically treated as comparison mistakes.

Version
7.22.0

## Checker
- Checker id: `category/java/errorprone.xml/UseEqualsToCompareStrings`
- Checker description: Detects string comparisons using '==' or '!=' operators instead of the equals method.

## Description of the false positive
`diff/output` marks 2 false positives for `category/java/errorprone.xml/UseEqualsToCompareStrings`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The recurring themes are string reference equality.

### `NegCase1.java`
Source: `output/checkstyle/StringLiteralEqualityCheck/src/main/java/scensct/core/neg/NegCase1.java`

The core operation is a reference-equality check involving strings. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Equality operator with string concatenation where first operand is not a string literal should not be flagged as having a string literal operand.

```java
// Equality operator with string concatenation where first operand is not a string literal should not be flagged as having a string literal operand.
package scensct.core.neg;

public class NegCase1 {
    public static void main(String[] args) {
        String dynamic = "dynamic";
        String result = dynamic + "suffix"; // Concatenation where first operand is not a literal
        String literal = "literal";
        boolean test = (result == literal); // Equality operator's direct operand is not a string literal
        System.out.println(test);
    }
}
```

### `NegCase3.java`
Source: `output/codeql/StringComparison/src/main/java/scensct/core/neg/NegCase3.java`

The core operation is a reference-equality check involving strings. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Reference equality test where both string operands are guaranteed interned should not be flagged as string reference equality violation.

```java
// Reference equality test where both string operands are guaranteed interned should not be flagged as string reference equality violation.
package scensct.core.neg;

public class NegCase3 {
    public static void main(String[] args) {
        String literal1 = "interned";
        String literal2 = "interned";
        
        // Scenario 3: both are string literals (interned)
        if (literal1 == literal2) {
            System.out.println("Both are interned literals");
        }
    }
}
```

## Cause analysis
The rule is documented as follows: Detects string comparisons using '==' or '!=' operators instead of the equals method.

In this set the problem shows up around string reference equality. The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `category/java/errorprone.xml/UseEqualsToCompareStrings`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
